src/plugins/typedispatcher/helpers.c

```diff
@@ -0,0 +1,437 @@
+	if (!tc) return NULL;
+	tc->type = SKEL;
+	tc->scope = keyNew (0, KEY_END);
+	tc->checks = ksNew (0, KS_END);
+	tc->types = ksNew (0, KS_END);
+	tc->params = ksNew (0, KS_END);
+	return tc;
+}
+
+// sets TypeConfig->type to SUMTYPE if keyString matches "SUM"
+// SUBTYPE otherwise
+void setTypeType (TypeConfig * tc, const Key * key)
+{
+	if (!strncasecmp (keyString (key), "SUM", 3))
+		tc->type = SUMTYPE;
+	else
+		tc->type = SUBTYPE;
+}
+
+TypeType getTypeType (TypeConfig * tc)
+{
+	return tc->type;
+}
+
+Key * getTypeKey (DispatchConfig * config, const char * type)
+{
+	Key * lookup = ksLookupByName (config->types, type, KDB_O_NONE);
+	return lookup;
+}
+
+// look up <type> and return its configuration if found
+TypeConfig * getType (DispatchConfig * config, const char * type)
+{
+	Key * lookup = getTypeKey (config, type);
+	if (!lookup) return NULL;
+	TypeConfig * tc = *(TypeConfig **)keyValue (lookup);
+	return tc;
+}
+
+char * replaceParametersWithArguments (const Key * key, KeySet * args)
+{
+	const char * ptr = keyString (key);
+#ifdef DEVBUILD
+	fprintf (stderr, "replacing parameters in %s\n", ptr);
+#endif
+	const char * lastPtr = ptr;
+	char * preparedString = NULL;
+	char * dstPtr = NULL;
+	ssize_t preparedStringLen = 1;
+	while (*ptr)
+	{
+		if (*ptr == '%')
+		{
+			const char * endPtr = ptr + 1;
+			if (lastPtr == ptr) ++lastPtr;
+			while (*endPtr)
+			{
+				if (*endPtr == '%')
+				{
+					// copy string from lastPtr to ptr
+					preparedStringLen += (ptr - lastPtr) + 1;
+					elektraRealloc ((void **)&preparedString, preparedStringLen);
+					if (!dstPtr)
+					{
+						dstPtr = preparedString;
+					}
+					else
+					{
+						dstPtr = preparedString + elektraStrLen (preparedString) - 1;
+					}
+					*dstPtr = '\0';
+					if (ptr >= lastPtr)
+					{
+						memcpy (dstPtr, lastPtr, (ptr - lastPtr));
+						dstPtr += (ptr - lastPtr);
+						*dstPtr = '\0';
+						lastPtr = endPtr + 1;
+					}
+
+
+					// extract argument name
+					ssize_t len = (endPtr - 1) - (ptr + 1) + 1; // length of argument string
+					char tmp[len + 1];			    // length + nullbyte
+					memset (tmp, 0, sizeof (tmp));
+					memcpy (tmp, ptr + 1, len);
+					Key * searchKey = keyNew (tmp, KEY_META_NAME, KEY_END);
+					Key * lookup = ksLookup (args, searchKey, KDB_O_NONE);
+					keyDel (searchKey);
+					if (lookup)
+					{
+
+						ssize_t newLen = preparedStringLen + elektraStrLen (keyString (lookup)) -
+								 1; // no need for 2 nullbytes
+						elektraRealloc ((void **)&preparedString, newLen);
+						strncat (preparedString, keyString (lookup), elektraStrLen (keyString (lookup)) - 1);
+						dstPtr = preparedString + newLen - 1;
+						preparedStringLen = newLen;
+					}
+					else
+					{
+						ssize_t newLen = elektraStrLen (preparedString) + sizeof (tmp) + 2;
+						elektraRealloc ((void **)&preparedString, newLen);
+						strcat (preparedString, "%");
+						strncat (preparedString, tmp, sizeof (tmp));
```

Comment:
The compiler gives me this warning. shouldn't be strlen be used instead of sizeof here?
helpers.c:274:37: warning: size argument in 'strncat'
      call appears to be size of the source [-Wstrncat-size]
                                                strncat (preparedString, tmp, sizeof (tmp));